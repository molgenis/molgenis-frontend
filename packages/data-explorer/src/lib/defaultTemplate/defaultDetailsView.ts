export const defaultDetailsView = `<div class="d-flex m-2">
<h2>{{ record.label || record.name }}</h2>
<div class="ml-auto" v-if="isShop"><shopping-cart-button :id="record.id" />
</div>
</div>
<ul class="list-group list-group-flush">
<li v-for="(value, propertyName, index) in record" :key="index" class="list-group-item border-0">
    <h5>{{ propertyName }}</h5>
    <template v-if="!isObject(value)">
        <p>{{ value }}</p>
    </template>
    <template v-else-if="isArray(value)">
        {{ value.map(v => v.label).join(', ') }}
    </template>
    <template v-else>
        <ul class="list-group list-group-flush">
            <li v-for="(refValue, refPropertyName, refIndex) in value" :key="refIndex" class="list-group-item border-0">
                <h5>{{ refPropertyName }}</h5>
                <p>{{ refValue }}</p>
            </li>
        </ul>
    </template>
</li>
</ul>`
