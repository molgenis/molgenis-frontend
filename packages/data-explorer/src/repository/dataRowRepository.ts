// @ts-ignore
import api from '@molgenis/molgenis-api-client'

const flattenCompounds = (fields: any[]) => {
  return fields.reduce((accum: any[], f: { type: string; children: any }) => {
    if (f.type === 'field-group') {
      accum = accum.concat(flattenCompounds(f.children))
    } else {
      accum.push(f)
    }
    return accum
  }, [])
}

const isFileIncluded = (formData: { [x: string]: any }, formFields: any) => {
  const flattendFields = flattenCompounds(formFields)
  const fieldsWithFile = flattendFields
    .filter((field: { type: string }) => field.type === 'file')
    .find((field: { id: string | number }) => typeof formData[field.id] !== 'string')

  return !!fieldsWithFile
}

export const appendToForm = (fields: any[], formData: FormData, [key, value]: [string, unknown]) => {
  const isFile = value && fields.find((field: { id: any; type: string }) => field.id === key && field.type === 'file' && typeof value !== 'string')
  if (isFile) {
    // @ts-ignore
    formData.append(key, value, value.name)
  } else {
    // @ts-ignore
    const stringValue: string = value === undefined || value === null ? '' : value
    formData.append(key, stringValue)
  }
}

const buildFormData = (data: { [s: string]: unknown } | ArrayLike<unknown>, fields: any) => {
  const formData = new FormData()
  const flattendFields = flattenCompounds(fields)
  Object.entries(data).forEach((pair) => appendToForm(flattendFields, formData, pair))
  return formData
}

const doPost = (uri: string, formData: any, formFields: any) => {
  const containsFileData = isFileIncluded(formData, formFields)
  const options = {
    headers: {
      'Accept': 'application/json',
      'X-Requested-With': 'XMLHttpRequest'
    },
    body: containsFileData ? buildFormData(formData, formFields) : JSON.stringify(formData),
    method: 'POST',
    credentials: 'same-origin'
  }

  return api.post(uri, options, containsFileData)
}

/**
 * Fetch for edit-response mixes data and metadata,
 * this function creates a new object with separate properties for data and metadata, similar to the
 * fetch for create response
 * @param response
 * @returns Promise{{meta: *, rowData: *}}
 */
const parseEditResponse = (response: { [x: string]: any; _meta: any }) => {
  const { _meta, ...rowData } = response
  return Promise.resolve({meta: _meta, rowData: rowData})
}

const fetchForCreate = (tableId: string) => api.get('/api/v2/' + tableId + '?num=0')

const fetchForUpdate = (tableId: string, rowId: string) => {
  return api.get('/api/v2/' + tableId + '/' + rowId).then((response: any) => parseEditResponse(response))
}

const create = (formData: any, formFields: any, tableId: string) => {
  return doPost('/api/v1/' + tableId + '?_method=PUT', formData, formFields)
}

const update = (formData: any, formFields: any, tableId: string, rowId: string) => {
  return doPost('/api/v1/' + tableId + '/' + rowId + '?_method=PUT', formData, formFields)
}

const save = async (formData: any, formFields: any, tableId: any, rowId: null | string) => {
  return rowId === null ? create(formData, formFields, tableId) : update(formData, formFields, tableId, rowId)
}

const fetch = (tableId: any, rowId: null | string) => rowId === null ? fetchForCreate(tableId) : fetchForUpdate(tableId, rowId)

const fetchOption = async (location: string) => api.get(location)

export {
  save, fetch, fetchOption
}