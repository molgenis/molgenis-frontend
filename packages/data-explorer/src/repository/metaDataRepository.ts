import client from '@/lib/client'
import { AxiosResponse } from 'axios'
import * as mapper from './metaDataResponseMapper'
import { ResponseEntityType } from '../types/EntityTypeV3'
import { MetaData } from '../types/MetaData'

const metaDataCache:{ [s: string]: MetaData } = {}
const metaDataCue: { [s: string]: Promise<AxiosResponse<ResponseEntityType>> } = {}

const mapAndStore = (entityId:string, response: AxiosResponse<ResponseEntityType>) => {
  const entityType = response.data
  const metadata = mapper.toMetaData(entityType)

  // Cache mapped response to improve speed
  metaDataCache[entityId] = metadata
  return metadata
}

const fetchMetaDataById = async (entityId: string): Promise<MetaData> => {
  if (metaDataCache[entityId]) {
    return metaDataCache[entityId]
  }
  if (metaDataCue[entityId]) {
    return metaDataCue[entityId].then((result) => {
      if (metaDataCache[entityId]) {
        return metaDataCache[entityId]
      }
      return mapAndStore(entityId, result)
    })
  }

  const response = client.get<ResponseEntityType>(`/api/metadata/${entityId}`, {
    params: {
      flattenAttributes: true
    }
  })

  metaDataCue[entityId] = response
  const resolved = await response
  return mapAndStore(entityId, resolved)
}

const fetchMetaDataByURL = async (url: string): Promise<MetaData> => {
  // @ts-ignore
  return fetchMetaDataById(url.split('/').pop())
}

export {
  fetchMetaDataById,
  fetchMetaDataByURL
}
