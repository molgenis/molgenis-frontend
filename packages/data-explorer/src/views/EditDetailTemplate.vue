<template>
  <div v-if="!loading" id="entity-detail-container" class="mg-mainview">
    <div class="ml-2">
      <nav aria-label="breadcrumb">
        <ol class="breadcrumb">
          <li class="breadcrumb-item" aria-current="page">
            <router-link id="headItemTooltipId" :to="{ name: 'de-view', params: { entity: entityType}, query: $route.query }">
              <b-tooltip
                v-if="tableMeta.description" placement="bottom"
                target="headItemTooltipId"
                triggers="hover"
              >
                {{ tableMeta.description }}
              </b-tooltip>
              {{ tableMeta.label }}
            </router-link>
          </li>
          <li class="breadcrumb-item" aria-current="page">
            <span>
              {{ record[tableMeta.idAttribute.name] }}
            </span>
          </li>
        </ol>
      </nav>
      <h1>
        Template editor
      </h1>
      <div class="my-3">
        <b-card no-body>
          <b-tabs pills card vertical>
            <!-- Record -->
            <b-tab title="Record properties" class="property-tab" active>
              <em class="font-weight-light">Example usage: record.id</em>
              <prop-render class="template-values mt-2" :object="record" />
            </b-tab>
            <!-- MetaData -->
            <b-tab title="Metadata properties" class="property-tab">
              <em class="font-weight-light">Example usage: metaData.id</em>
              <prop-render class="template-values mt-2" :object="tableMeta" />
            </b-tab>
            <!-- Components -->
            <b-tab title="Components" class="property-tab">
              <table class="table table-striped">
                <thead>
                  <th>
                    Component:
                  </th>
                  <th>Code:</th>
                </thead>
                <tbody>
                  <tr>
                    <td>Shoppingcart Button</td>
                    <td>&lt;shopping-cart-button :id="record.id" /&gt;</td>
                  </tr>
                </tbody>
              </table>
            </b-tab>
            <b-tab title="Editor options">
              <button class="btn btn-outline-primary" @click="landscapeMode = !landscapeMode">
                Set {{ landscapeMode ? "Portrait" : "Landscape" }}
              </button>
            </b-tab>
          </b-tabs>
        </b-card>
      </div>
      <div 
        class="d-flex w-100 flex-wrap"
        :class="{ 'flex-column': landscapeMode }"
      >
        <div 
          class="mr-3 mt-3"
          :class="[{'w-100': landscapeMode}, {'w-50': !landscapeMode}]"
        >
          <!-- Editor -->
          <span class="font-weight-bold mb-1">Template</span>
          <form-component
            id="template"
            class="border-top"
            :options="{ showEyeButton: false }"
            :form-fields="formFields"
            :initial-form-data="{ template: template }"
            @valueChange="template = $event.template"
          />
          <button 
            class="btn btn-primary mb-2 ml-2" 
            type="submit" 
            :disabled="isSavingTemplate"
            @click.prevent="saveTemplate"
          >
            Save template <font-awesome-icon v-if="isSavingTemplate" icon="spinner" spin />
          </button>
        </div>
        <div
          class="preview-container mt-3"
          :class="[{'w-100': landscapeMode}, {'ml-auto': !landscapeMode}]"
        >
          <span class="font-weight-bold mb-1">Preview</span>
          <div v-if="template" class="template-preview border">
            <custom-entity-detail
              :record="record"
              :meta-data="tableMeta" :template="template"
            />
          </div>
          <div v-else class="template-preview p-2 font-italic text-muted">
            No template set for current entity
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { defaultDetailsView } from '@/lib/defaultTemplate/defaultDetailsView'
import { getRowDataWithReferenceLabels } from '@/repository/dataRepository'
import { mapActions, mapState } from 'vuex'
import CustomEntityDetail from '@/components/dataView/CustomEntityDetail.vue'
import PropRender from '@/components/dataView/PropRender.vue'
import { FormComponent } from '@molgenis/molgenis-ui-form'

export default {
  name: 'EditDetailTemplate',
  components: { CustomEntityDetail, PropRender, FormComponent },
  props: {
    entityType: {
      type: String,
      required: true
    },
    entity: {
      type: String,
      required: true
    }
  },
  data () {
    return {
      landscapeMode: false,
      currentTemplate: '',
      loading: true,
      isSavingTemplate: false,
      record: null,
      formFields: [
        {
          id: 'template',
          label: 'Template',
          description: 'Edit the template to change the layout, use record.[property-name] and metaData.[property-name] to refer to the entity (meta)data.',
          type: 'script',
          visible: () => true,
          required: () => false,
          validate: () => true
        }
      ],
      formState: {}
    }
  },
  computed: {
    ...mapState('explorer', ['tableSettings', 'tableMeta']),
    template: {
      get () {
        return this.tableSettings.customDetailCode ? this.tableSettings.customDetailCode : this.baseTemplate()
      },
      set (value) {
        this.tableSettings.customDetailCode = value
      }
    }
  },
  async mounted () {
    this.fetchTablePermissions({ tableName: this.entityType })
    await this.fetchTableMeta({ tableName: this.entityType })
    // Wait for both items to be present, but only the second item is needed.
    const [, data] = await Promise.all([
      this.fetchTableSettings({ tableName: this.entityType }),
      getRowDataWithReferenceLabels(this.entityType, this.entity, this.tableMeta)
    ])
    this.record = data
    this.loading = false
  },
  methods: {
    ...mapActions('explorer', [
      'fetchTablePermissions',
      'fetchTableMeta',
      'fetchTableSettings',
      'saveEntityDetailTemplate'
    ]),
    saveTemplate: async function () {
      this.isSavingTemplate = true
      await this.saveEntityDetailTemplate({ template: this.template })
      this.$router.push({ name: 'entity-detail', params: { entityType: this.entityType, entity: this.entity } })
    },
    baseTemplate() {
      return defaultDetailsView
    }
  },
}
</script>

<style scoped>

  .preview-container {
    min-width: 42rem;
    width: 48%;
  }

  .template-preview {
    max-height: 75vh;
    overflow: auto;
  }

  .template-values {
    font-family: 'Courier New', monospace;
  }

  .property-tab {
    height: 200px;
    max-height: 200px;
    overflow: auto;
  }

  /* 
    ::v-deep is the stylelint accepted version, does the same as >>> and /deep/ 
    >>> is the deep selector for plain css (if you don't use stylelint)
    /deep/ is the scss selector
  */

  ::v-deep label[for='template'] {
    display: none;
  }

  ::v-deep .CodeMirror {
    height: 25.25rem !important;
  }

  ::v-deep #template-description {
    margin-left: 1rem;
  }
</style>
