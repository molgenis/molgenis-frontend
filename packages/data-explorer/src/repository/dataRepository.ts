import { buildExpandedAttributesQuery } from './queryBuilder'
import { getAttributesfromMeta } from './metaDataService'
import * as metaDataRepository from './metaDataRepository'
import { DataApiResponse, DataApiResponseItem, DataObject } from '@/types/ApiResponse'
import { MetaData, Attribute } from '@/types/MetaData'
import client from '@/lib/client'
import { encodeRsqlValue } from '@molgenis/rsql'
import { Pagination } from '@molgenis-ui/components-library'

// maps api response to object with as key the name of the column and as value the label of the value or a list of labels for mrefs
const levelOneRowMapper = async (rowData: DataApiResponseItem, metaData: MetaData) => {
  if (!rowData.data) {
    throw new Error('invalid input: row data must have data property')
  }

  const metadataAttributeMap: {[s: string]: Attribute} = metaData.attributes.reduce((accum:any, attr) => {
    accum[attr.name] = attr
    return accum
  }, {})

  const row: DataObject = rowData.data
  let result:any = {}
  let keys = Object.keys(row)

  for (let i = 0; i < keys.length; i++) {
    const columnKey = keys[i]
    const value = row[columnKey]
    const attributeMetadata = metadataAttributeMap[columnKey]
    const isReference = attributeMetadata.isReference
    let resolvedValue
    if (isReference && attributeMetadata.refEntityType) {
      // @ts-ignore
      if (value.items) {
        resolvedValue = await getMrefLabel(attributeMetadata, value)
      } else {
        resolvedValue = await getRefLabel(attributeMetadata, value)
      }
    } else {
      resolvedValue = value
    }
    result[columnKey] = resolvedValue
  }

  return result
}

const getMrefLabel = async (attributeMetadata: Attribute, rowDataItem:any) => {
  // The isMref already checks if the value.items is available
  // @ts-ignore
  const refMetadata = await metaDataRepository.fetchMetaDataByURL(attributeMetadata.refEntityType)
  // @ts-ignore
  const resolvedValue = rowDataItem.items.map((mrefValue) => mrefValue.data[refMetadata.labelAttribute.name]).join(', ')
  return resolvedValue
}

const getRefLabel = async (attributeMetadata: Attribute, rowDataItem:any) => {
  // This is checked by isReference
  // @ts-ignore
  const refMetadata = await metaDataRepository.fetchMetaDataByURL(attributeMetadata.refEntityType)
  // @ts-ignore
  const resolvedValue = rowDataItem.data[refMetadata.labelAttribute.name]
  return resolvedValue
}
const apiResponseMapper = (rowData: DataApiResponseItem) => {
  if (rowData.data) {
    const row: any = rowData.data
    return Object.keys(row).reduce((accum: { [s: string]: string | object }, key) => {
      // check if ref
      if (typeof row[key] === 'object') {
        // This is checked by the line above
        // @ts-ignore
        accum[key] = levelNRowMapper(row[key])
      } else {
        accum[key] = row[key]
      }
      return accum
    }, {})
  }
}

const levelNRowMapper = (rowData: DataApiResponseItem) => {
  // check if mref
  if (rowData.items) {
    const rows: DataApiResponseItem[] = rowData.items
    return rows.map((rowData) => {
      return apiResponseMapper(rowData)
    })
  } else {
    return apiResponseMapper(rowData)
  }
}

const addFilterIfSet = (request: string, rsqlFilter?: string): string => {
  return rsqlFilter ? `${request}&q=${encodeRsqlValue(rsqlFilter)}` : request
}

const pageQuery = (pagination :Pagination|undefined) => {
  // (!) Adapt page to 0-based data-api page property.
  return pagination ? Object
    .entries({ page: pagination.page - 1, size: pagination.size })
    .map(([k, v]) => `${k}=${v}`)
    .join('&') + '&'
    : ''
}

const getTableDataDeepReference = async (
  tableId: string,
  metaData: MetaData,
  coloms: string[],
  rsqlQuery?: string,
  pagination?: Pagination
) => {
  if (!coloms.includes(metaData.idAttribute.name)) {
    coloms.push(metaData.idAttribute.name)
  }

  if (metaData.labelAttribute !== undefined && !coloms.includes(metaData.labelAttribute.name)) {
    coloms.push(metaData.labelAttribute.name)
  }

  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, coloms)

  const request = addFilterIfSet(`/api/data/${tableId}?${pageQuery(pagination)}${expandReferencesQuery}`, rsqlQuery)
  const response = (await client.get<DataApiResponse>(request)).data

  const items = response.items.map((item: DataApiResponseItem) => levelNRowMapper(item))
  return { items, page: response.page }
}

const getTableDataWithLabel = async (tableId: string, metaData: MetaData, columns: string[], rsqlQuery?: string, pagination?: Pagination) => {
  const columnSet = new Set([...columns])
  columnSet.add(metaData.idAttribute.name)
  if (metaData.labelAttribute !== undefined) {
    columnSet.add(metaData.labelAttribute.name)
  }

  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, [...columnSet])
  const request = addFilterIfSet(`/api/data/${tableId}?${pageQuery(pagination)}${expandReferencesQuery}`, rsqlQuery)
  const response = (await client.get<DataApiResponse>(request)).data
  // @ts-ignore
  const items = await Promise.all(response.items.map(async (item: DataApiResponseItem) => {
    return levelOneRowMapper(item, metaData)
  }))

  return { items, page: response.page }
}

// called on row expand
const getRowDataWithReferenceLabels = async (tableId: string, rowId: string, metaData: MetaData, pagination?: Pagination) => {
  const attributes: string[] = getAttributesfromMeta(metaData)
  // Todo: remove work around, needed as compounds are not passed by getAttributesfromMeta.
  // Adding id and label makes sure we get these fields.
  const columnSet = new Set([...attributes])
  columnSet.add(metaData.idAttribute.name)
  if (metaData.labelAttribute !== undefined) {
    columnSet.add(metaData.labelAttribute.name)
  }
  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, [...columnSet])
  const response = await client.get<DataApiResponse>(`/api/data/${tableId}/${rowId}?${pageQuery(pagination)}${expandReferencesQuery}`)
  return levelOneRowMapper(response.data, metaData)
}

const deleteRow = async (tableId: string, rowId: string) => {
  return client.delete(`/api/data/${tableId}/${rowId}`)
}

export {
  getTableDataDeepReference,
  getTableDataWithLabel,
  getRowDataWithReferenceLabels,
  deleteRow
}
