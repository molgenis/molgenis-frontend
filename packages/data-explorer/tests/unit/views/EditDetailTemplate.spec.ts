import { createLocalVue, shallowMount } from '@vue/test-utils'
import EditDetailTemplate from '@/views/EditDetailTemplate.vue'
import { getRowDataWithReferenceLabels } from '@/repository/dataRepository'
import { defaultDetailsView } from '@/lib/defaultTemplate/defaultDetailsView'
import Vuex from 'vuex'


jest.mock('@/repository/dataRepository', () => {
  return {
    getRowDataWithReferenceLabels: jest.fn()
  }
})

const mocks = {
  $bvModal: { msgBoxConfirm: jest.fn() },
  $router: { push: jest.fn() },
  $route: { query: { foo: 'bar' } }
}
const stubs = ['font-awesome-icon', 'router-link', 'b-tooltip', 'b-tab', 'b-tabs', 'b-card']
const directives = { 'b-tooltip': () => {} }
let actions: any
let getters: any
let state: any
const localVue = createLocalVue()
localVue.use(Vuex)

describe('EditDetailTemplate.vue', () => {
  let wrapper
  beforeEach(async () => {

    // @ts-ignore
    getRowDataWithReferenceLabels.mockResolvedValue({
      id: 'abc-123',
      key1: 'val1',
      key2: ['val2', 'val3'],
      key3: {
        key4: 'val'
      }
    })

    actions = {
      fetchTablePermissions: jest.fn(),
      fetchTableSettings: jest.fn(),
      fetchTableMeta: jest.fn(),
      saveEntityDetailTemplate: jest.fn()
    }

    getters = {
      hasAddRights: () => true,
      hasEditRights: () => true,
      hasDeleteRights: () => true,
      hasEditSettingsRights: () => true
    }

    state = {
      tableMeta: {
        label: 'my label',
        description: 'my description',
        idAttribute: {
          name: 'id'
        }
      } ,
      tableSettings: {
        customDetailCode: '<h1>I m not a template</h1>'
      }
    }

    const explorer = { state, actions, getters, namespaced: true }
    const store = new Vuex.Store({ modules: { explorer } })
    const propsData = { entityType: 'entityType', entity: 'entity' }

    wrapper = shallowMount(EditDetailTemplate, { propsData, store, localVue, stubs, mocks, directives })
    await wrapper.vm.$nextTick // meta
    await wrapper.vm.$nextTick // settings
    await wrapper.vm.$nextTick // data
  })

  afterAll(() => {
    wrapper.destroy()
  })

  it('should render the editor', () => {
    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when the template is edited', () => {
    beforeEach(async() => {
      await wrapper.setData({ template: '<p>edit</p>' })
    })

    it('should update the template', () => {
      expect(wrapper.vm.template).toEqual('<p>edit</p>')
    })
  })

  describe('when the template is saved', () => {
    beforeEach(async() => {
      await wrapper.setData({ template: '<p>edit</p>' })
      await wrapper.find('button[type=submit]').trigger('click')
    })

    it('should update the template', async () => {
      expect(actions.saveEntityDetailTemplate).toHaveBeenCalledWith(expect.anything(), { template: "<p>edit</p>" } )
    })
  })

  describe('when no template is set', () => {
    beforeEach(async() => {
      state = {
        tableMeta: {
          label: 'my label',
          description: 'my description',
          idAttribute: {
            name: 'id'
          }
        } ,
        tableSettings: {
          customDetailCode: null
        }
      }
      const explorer = { state, actions, getters, namespaced: true }
      const store = new Vuex.Store({ modules: { explorer } })
      const propsData = { entityType: 'entityType', entity: 'entity' }
      wrapper = shallowMount(EditDetailTemplate, { propsData, store, localVue, stubs, mocks, directives })
    })

    it('should render the default template if no custom code is specified', () => {
      expect(wrapper.vm.template).toEqual(defaultDetailsView)
    })
  })

})
