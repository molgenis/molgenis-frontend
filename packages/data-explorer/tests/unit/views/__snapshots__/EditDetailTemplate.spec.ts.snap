// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDetailTemplate.vue should render the editor 1`] = `
<div id="entity-detail-container" class="mg-mainview">
  <div class="ml-2">
    <nav aria-label="breadcrumb">
      <ol class="breadcrumb">
        <li aria-current="page" class="breadcrumb-item">
          <router-link-stub id="headItemTooltipId" to="[object Object]">
            <b-tooltip-stub placement="bottom" target="headItemTooltipId" triggers="hover">
              my description
            </b-tooltip-stub>
            my label
          </router-link-stub>
        </li>
        <li aria-current="page" class="breadcrumb-item"><span>
            abc-123
          </span></li>
      </ol>
    </nav>
    <h1>
      Template editor
    </h1>
    <div class="my-3">
      <b-card-stub no-body="">
        <b-tabs-stub pills="" card="" vertical="">
          <b-tab-stub title="Record properties" active="" class="property-tab"><em class="font-weight-light">Example usage: record.id</em>
            <prop-render-stub object="[object Object]" class="template-values mt-2"></prop-render-stub>
          </b-tab-stub>
          <b-tab-stub title="Metadata properties" class="property-tab"><em class="font-weight-light">Example usage: metaData.id</em>
            <prop-render-stub object="[object Object]" class="template-values mt-2"></prop-render-stub>
          </b-tab-stub>
          <b-tab-stub title="Components" class="property-tab">
            <table class="table table-striped">
              <thead>
                <th>
                  Component:
                </th>
                <th>Code:</th>
              </thead>
              <tbody>
                <tr>
                  <td>Shoppingcart Button</td>
                  <td>&lt;shopping-cart-button :id="record.id" /&gt;</td>
                </tr>
              </tbody>
            </table>
          </b-tab-stub>
          <b-tab-stub title="Editor options"><button class="btn btn-outline-primary">
              Set Landscape
            </button></b-tab-stub>
        </b-tabs-stub>
      </b-card-stub>
    </div>
    <div class="d-flex w-100 flex-wrap">
      <div class="mr-3 mt-3 w-50"><span class="font-weight-bold mb-1">Template</span>
        <form-component-stub id="template" formfields="[object Object]" initialformdata="[object Object]" formstate="[object Object]" options="[object Object]" class="border-top"></form-component-stub> <button type="submit" class="btn btn-primary mb-2 ml-2">
          Save template
          <!---->
        </button>
      </div>
      <div class="preview-container mt-3 ml-auto"><span class="font-weight-bold mb-1">Preview</span>
        <div class="template-preview border">
          <custom-entity-detail-stub record="[object Object]" metadata="[object Object]" template="<h1>I m not a template</h1>"></custom-entity-detail-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;
